# Course Recommedation system

### Read Data

In [1]:
import numpy as np
import pandas as pd

In [2]:
courses = pd.read_csv('Vishwa_scrape_course_dataset(3000).csv')

In [3]:
courses

id                                              title  \
0     c00001  The Complete Python Bootcamp From Zero to Hero...   
1     c00002         The Complete 2023 Web Development Bootcamp   
2     c00003                    The Web Developer Bootcamp 2023   
3     c00004  100 Days of Code: The Complete Python Pro Boot...   
4     c00005  React - The Complete Guide 2023 (incl. React R...   
...      ...                                                ...   
3003  c03004  Complete JS Bootcamp | JavaScript Programming ...   
3004  c03005  Computer Vision: Face Recognition Quick Starte...   
3005  c03006             Statistics with R - Intermediate Level   
3006  c03007  Anime Character Creator: Make 3D Anime Charact...   
3007  c03008  Java Certification : OCP (1Z0-809) Exam Simula...   

                                                   desc  \
0     Learn Python like a Professional  Start from t...   
1     Become a Full-Stack Web Developer with just ON...   
2     10 Hours of React just added. Become a Develop...   
3     Master Python by building 100 projects in 100 ...   
4     Dive in and learn React.js from scratch! Learn...   
...                                                 ...   
3003  Complete JS Bootcamp | JavaScript Programming ...   
3004  Python Deep Learning based Face Detection, Rec...   
3005           Statistical analyses using the R program   
3006  Model, rig and pose an anime character in this...   
3007  Pass the Oracle Certified Professional(OCP): J...   

                                         by  
0                             Jose Portilla  
1                             Dr. Angela Yu  
2                               Colt Steele  
3                             Dr. Angela Yu  
4     Academind by Maximilian Schwarzmüller  
...                                     ...  
3003                           Faisal Zamir  
3004                        Abhilash Nelson  
3005                      Bogdan Anastasiei  
3006                           Grant Abbitt  
3007                         Udayan Khattry  

[3008 rows x 4 columns]

### Remove null values

In [4]:
courses.isnull().sum()

id       0
title    0
desc     1
by       0
dtype: int64

In [5]:
courses.dropna(inplace=True)

In [6]:
courses.isnull().sum()

id       0
title    0
desc     0
by       0
dtype: int64

### index reseting is imprtant to avoid further confusion (it turns 0 to 3007 to 0 to 3006)

In [7]:
courses = courses.reset_index()

In [8]:
courses

index      id                                              title  \
0         0  c00001  The Complete Python Bootcamp From Zero to Hero...   
1         1  c00002         The Complete 2023 Web Development Bootcamp   
2         2  c00003                    The Web Developer Bootcamp 2023   
3         3  c00004  100 Days of Code: The Complete Python Pro Boot...   
4         4  c00005  React - The Complete Guide 2023 (incl. React R...   
...     ...     ...                                                ...   
3002   3003  c03004  Complete JS Bootcamp | JavaScript Programming ...   
3003   3004  c03005  Computer Vision: Face Recognition Quick Starte...   
3004   3005  c03006             Statistics with R - Intermediate Level   
3005   3006  c03007  Anime Character Creator: Make 3D Anime Charact...   
3006   3007  c03008  Java Certification : OCP (1Z0-809) Exam Simula...   

                                                   desc  \
0     Learn Python like a Professional  Start from t...   
1     Become a Full-Stack Web Developer with just ON...   
2     10 Hours of React just added. Become a Develop...   
3     Master Python by building 100 projects in 100 ...   
4     Dive in and learn React.js from scratch! Learn...   
...                                                 ...   
3002  Complete JS Bootcamp | JavaScript Programming ...   
3003  Python Deep Learning based Face Detection, Rec...   
3004           Statistical analyses using the R program   
3005  Model, rig and pose an anime character in this...   
3006  Pass the Oracle Certified Professional(OCP): J...   

                                         by  
0                             Jose Portilla  
1                             Dr. Angela Yu  
2                               Colt Steele  
3                             Dr. Angela Yu  
4     Academind by Maximilian Schwarzmüller  
...                                     ...  
3002                           Faisal Zamir  
3003                        Abhilash Nelson  
3004                      Bogdan Anastasiei  
3005                           Grant Abbitt  
3006                         Udayan Khattry  

[3007 rows x 5 columns]

In [9]:
courses[1750:1780]

index      id                                              title  \
1750   1750  c01751  Reactive Programming in Modern Java using Proj...   
1751   1751  c01752  The Expert’s Guide of Selenium with Python | B...   
1752   1752  c01753            The Full Stack Web Development Bootcamp   
1753   1753  c01754  Azure Machine Learning &  MLOps : Beginner to ...   
1754   1754  c01755  Object Oriented Programming (OOP) in PHP - Bui...   
1755   1755  c01756    REST Assured. API test automation for beginners   
1756   1756  c01757  Build an Amazon Affiliate E-Commerce Store fro...   
1757   1757  c01758                     Remix.js - The Practical Guide   
1758   1758  c01759  AWS Lambda - A Practical Guide - Learn from an...   
1759   1759  c01760                 QR Code Generator Using JavaScript   
1760   1760  c01761                APIs in PHP: from Basic to Advanced   
1761   1762  c01763    The Complete Guide to Bubble Visual Programming   
1762   1763  c01764  Java for Complete Beginners: Become Junior Jav...   
1763   1764  c01765  Learn Advanced T-SQL For SQL Server Administrator   
1764   1765  c01766        Mastering Regular Expressions in JavaScript   
1765   1766  c01767  Build Creative Website Using HTML5, CSS3, jQue...   
1766   1767  c01768  Mastering Microsoft Power Apps 2023: From Zero...   
1767   1768  c01769                           Learn Redis from Scratch   
1768   1769  c01770                  Math for Data Science Masterclass   
1769   1770  c01771      DevOps Fundamentals: Gain Solid Understanding   
1770   1771  c01772                               Golang for Beginners   
1771   1772  c01773      Break Away: Programming And Coding Interviews   
1772   1773  c01774  Python 3: Project-based Python, Algorithms, Da...   
1773   1774  c01775  Complete Python Bootcamp For Everyone From Zer...   
1774   1775  c01776    Write Asynchronous C# Code With Tasks and PLINQ   
1775   1776  c01777  Build a Stock Market Web App With Python and D...   
1776   1777  c01778  Recursion, Backtracking and Dynamic Programmin...   
1777   1778  c01779               C++ Programming for Beginners (2023)   
1778   1779  c01780            Working with Concurrency in Go (Golang)   
1779   1780  c01781     Intro to Game Development using Unity - Part I   

                                                   desc  \
1750  Learn to write fast performing Asynchronous an...   
1751  Pytest | BDD Framework using Behave | Robot Fr...   
1752  Learn HTML, CSS, JavaScript, jQuery, PHP & myS...   
1753  The most compressive  course covering MLOps an...   
1754  No PHP knowledge required. Everything you need...   
1755  Rest API test automation from scratch. Create ...   
1756  Earn Passive Income by Building an Amazon Affi...   
1757  Learn how to build feature-rich & highly react...   
1758  Learn the fundamentals of AWS Lambda and 3 pra...   
1759  Learn How To Generate QR Code Using JavaScript...   
1760  Use REST APIs from PHP, and create your own RE...   
1761  Learn how to build sophisticated web applicati...   
1762  Java programming for complete beginners. Creat...   
1763            Getting to understand stored procedures   
1764  Full of Opportunities to Learn by Doing in Ord...   
1765  Learn HTML5, CSS3,  jQuery, Bootstrap And Buil...   
1766  Unlock the Power of Low-Code with MS PowerApps...   
1767  A comprehensive guide to learn and use the Red...   
1768  Learn about probability, statistics, and more ...   
1769  Learn the fundamentals of DevOps including the...   
1770  Learn the Go programming language with Hands-O...   
1771  A course that teaches pointers, linked lists, ...   
1772  Learn programming with Python 3; visualize Alg...   
1773  Master Python Programming by building 100+ REA...   
1774  A comprehensive course teaching you how to wri...   
1775  Learn Database Driven Web Development With Dja...   
1776  Learn Competitive Programming, Recursion, Back...   
1777  Enter the world of computer science, learn to ...   


In [10]:
courses = courses[['id','title','desc','by']]

In [11]:
courses

id                                              title  \
0     c00001  The Complete Python Bootcamp From Zero to Hero...   
1     c00002         The Complete 2023 Web Development Bootcamp   
2     c00003                    The Web Developer Bootcamp 2023   
3     c00004  100 Days of Code: The Complete Python Pro Boot...   
4     c00005  React - The Complete Guide 2023 (incl. React R...   
...      ...                                                ...   
3002  c03004  Complete JS Bootcamp | JavaScript Programming ...   
3003  c03005  Computer Vision: Face Recognition Quick Starte...   
3004  c03006             Statistics with R - Intermediate Level   
3005  c03007  Anime Character Creator: Make 3D Anime Charact...   
3006  c03008  Java Certification : OCP (1Z0-809) Exam Simula...   

                                                   desc  \
0     Learn Python like a Professional  Start from t...   
1     Become a Full-Stack Web Developer with just ON...   
2     10 Hours of React just added. Become a Develop...   
3     Master Python by building 100 projects in 100 ...   
4     Dive in and learn React.js from scratch! Learn...   
...                                                 ...   
3002  Complete JS Bootcamp | JavaScript Programming ...   
3003  Python Deep Learning based Face Detection, Rec...   
3004           Statistical analyses using the R program   
3005  Model, rig and pose an anime character in this...   
3006  Pass the Oracle Certified Professional(OCP): J...   

                                         by  
0                             Jose Portilla  
1                             Dr. Angela Yu  
2                               Colt Steele  
3                             Dr. Angela Yu  
4     Academind by Maximilian Schwarzmüller  
...                                     ...  
3002                           Faisal Zamir  
3003                        Abhilash Nelson  
3004                      Bogdan Anastasiei  
3005                           Grant Abbitt  
3006                         Udayan Khattry  

[3007 rows x 4 columns]

In [12]:
type(courses.iloc[0].desc)

str

In [13]:
#Removing spaces from words to make it as a sinlge tag (i.e Jose Portilla is treated as two tags But JosePortilla is treated as one tag)
# courses['by'].apply(lambda x:[i.replace(" ","") for i in x])

#Notworking for by column

In [14]:
#function to remove all odd characters
import re

def clean_title(title):
    return re.sub("[^a-zA-Z0-9 ]","",title)

In [15]:
courses['desc'] = courses['desc'].apply(clean_title)
courses['by'] = courses['by'].apply(clean_title)

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_22320\2196737410.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  courses['desc'] = courses['desc'].apply(clean_title)
C:\Users\ADMIN\AppData\Local\Temp\ipykernel_22320\2196737410.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  courses['by'] = courses['by'].apply(clean_title)


In [16]:
courses.head(10)

id                                              title  \
0  c00001  The Complete Python Bootcamp From Zero to Hero...   
1  c00002         The Complete 2023 Web Development Bootcamp   
2  c00003                    The Web Developer Bootcamp 2023   
3  c00004  100 Days of Code: The Complete Python Pro Boot...   
4  c00005  React - The Complete Guide 2023 (incl. React R...   
5  c00006  The Complete JavaScript Course 2024: From Zero...   
6  c00007        Angular - The Complete Guide (2023 Edition)   
7  c00008          Java 17 Masterclass: Start Coding in 2023   
8  c00009  Machine Learning A-Z: AI, Python & R + ChatGPT...   
9  c00010  Python for Data Science and Machine Learning B...   

                                                desc  \
0  Learn Python like a Professional  Start from t...   
1  Become a FullStack Web Developer with just ONE...   
2  10 Hours of React just added Become a Develope...   
3  Master Python by building 100 projects in 100 ...   
4  Dive in and learn Reactjs from scratch Learn R...   
5  The modern JavaScript course for everyone Mast...   
6  Master Angular formerly Angular 2 and build aw...   
7  Acquire Key Java Skills From Basics to Advance...   
8  Learn to create Machine Learning Algorithms in...   
9  Learn how to use NumPy Pandas Seaborn  Matplot...   

                                     by  
0                         Jose Portilla  
1                          Dr Angela Yu  
2                           Colt Steele  
3                          Dr Angela Yu  
4  Academind by Maximilian Schwarzmller  
5                     Jonas Schmedtmann  
6               Maximilian Schwarzmller  
7                          Tim Buchalka  
8                       Kirill Eremenko  
9                         Jose Portilla

In [17]:
#Convert string(sentence) to lists
courses['desc'] =  courses['desc'].apply(lambda x:x.split())

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_22320\2602616792.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  courses['desc'] =  courses['desc'].apply(lambda x:x.split())


In [18]:
courses['by'] =  courses['by'].apply(lambda x:x.split())

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_22320\4198081956.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  courses['by'] =  courses['by'].apply(lambda x:x.split())


In [19]:
courses['temp_title'] = courses['title']

In [20]:
courses['temp_title'] = courses['temp_title'].apply(clean_title)

In [21]:
courses['temp_title'] =  courses['temp_title'].apply(lambda x:x.split())

In [22]:
courses.head(10)

id                                              title  \
0  c00001  The Complete Python Bootcamp From Zero to Hero...   
1  c00002         The Complete 2023 Web Development Bootcamp   
2  c00003                    The Web Developer Bootcamp 2023   
3  c00004  100 Days of Code: The Complete Python Pro Boot...   
4  c00005  React - The Complete Guide 2023 (incl. React R...   
5  c00006  The Complete JavaScript Course 2024: From Zero...   
6  c00007        Angular - The Complete Guide (2023 Edition)   
7  c00008          Java 17 Masterclass: Start Coding in 2023   
8  c00009  Machine Learning A-Z: AI, Python & R + ChatGPT...   
9  c00010  Python for Data Science and Machine Learning B...   

                                                desc  \
0  [Learn, Python, like, a, Professional, Start, ...   
1  [Become, a, FullStack, Web, Developer, with, j...   
2  [10, Hours, of, React, just, added, Become, a,...   
3  [Master, Python, by, building, 100, projects, ...   
4  [Dive, in, and, learn, Reactjs, from, scratch,...   
5  [The, modern, JavaScript, course, for, everyon...   
6  [Master, Angular, formerly, Angular, 2, and, b...   
7  [Acquire, Key, Java, Skills, From, Basics, to,...   
8  [Learn, to, create, Machine, Learning, Algorit...   
9  [Learn, how, to, use, NumPy, Pandas, Seaborn, ...   

                                          by  \
0                           [Jose, Portilla]   
1                           [Dr, Angela, Yu]   
2                             [Colt, Steele]   
3                           [Dr, Angela, Yu]   
4  [Academind, by, Maximilian, Schwarzmller]   
5                       [Jonas, Schmedtmann]   
6                 [Maximilian, Schwarzmller]   
7                            [Tim, Buchalka]   
8                         [Kirill, Eremenko]   
9                           [Jose, Portilla]   

                                          temp_title  
0  [The, Complete, Python, Bootcamp, From, Zero, ...  
1  [The, Complete, 2023, Web, Development, Bootcamp]  
2              [The, Web, Developer, Bootcamp, 2023]  
3  [100, Days, of, Code, The, Complete, Python, P...  
4  [React, The, Complete, Guide, 2023, incl, Reac...  
5  [The, Complete, JavaScript, Course, 2024, From...  
6     [Angular, The, Complete, Guide, 2023, Edition]  
7   [Java, 17, Masterclass, Start, Coding, in, 2023]  
8  [Machine, Learning, AZ, AI, Python, R, ChatGPT...  
9  [Python, for, Data, Science, and, Machine, Lea...

In [23]:
#the ultimate concatenation
courses['tags'] = courses['temp_title'] + courses['desc'] + courses['by'] 

In [24]:
courses.head(10)

id                                              title  \
0  c00001  The Complete Python Bootcamp From Zero to Hero...   
1  c00002         The Complete 2023 Web Development Bootcamp   
2  c00003                    The Web Developer Bootcamp 2023   
3  c00004  100 Days of Code: The Complete Python Pro Boot...   
4  c00005  React - The Complete Guide 2023 (incl. React R...   
5  c00006  The Complete JavaScript Course 2024: From Zero...   
6  c00007        Angular - The Complete Guide (2023 Edition)   
7  c00008          Java 17 Masterclass: Start Coding in 2023   
8  c00009  Machine Learning A-Z: AI, Python & R + ChatGPT...   
9  c00010  Python for Data Science and Machine Learning B...   

                                                desc  \
0  [Learn, Python, like, a, Professional, Start, ...   
1  [Become, a, FullStack, Web, Developer, with, j...   
2  [10, Hours, of, React, just, added, Become, a,...   
3  [Master, Python, by, building, 100, projects, ...   
4  [Dive, in, and, learn, Reactjs, from, scratch,...   
5  [The, modern, JavaScript, course, for, everyon...   
6  [Master, Angular, formerly, Angular, 2, and, b...   
7  [Acquire, Key, Java, Skills, From, Basics, to,...   
8  [Learn, to, create, Machine, Learning, Algorit...   
9  [Learn, how, to, use, NumPy, Pandas, Seaborn, ...   

                                          by  \
0                           [Jose, Portilla]   
1                           [Dr, Angela, Yu]   
2                             [Colt, Steele]   
3                           [Dr, Angela, Yu]   
4  [Academind, by, Maximilian, Schwarzmller]   
5                       [Jonas, Schmedtmann]   
6                 [Maximilian, Schwarzmller]   
7                            [Tim, Buchalka]   
8                         [Kirill, Eremenko]   
9                           [Jose, Portilla]   

                                          temp_title  \
0  [The, Complete, Python, Bootcamp, From, Zero, ...   
1  [The, Complete, 2023, Web, Development, Bootcamp]   
2              [The, Web, Developer, Bootcamp, 2023]   
3  [100, Days, of, Code, The, Complete, Python, P...   
4  [React, The, Complete, Guide, 2023, incl, Reac...   
5  [The, Complete, JavaScript, Course, 2024, From...   
6     [Angular, The, Complete, Guide, 2023, Edition]   
7   [Java, 17, Masterclass, Start, Coding, in, 2023]   
8  [Machine, Learning, AZ, AI, Python, R, ChatGPT...   
9  [Python, for, Data, Science, and, Machine, Lea...   

                                                tags  
0  [The, Complete, Python, Bootcamp, From, Zero, ...  
1  [The, Complete, 2023, Web, Development, Bootca...  
2  [The, Web, Developer, Bootcamp, 2023, 10, Hour...  
3  [100, Days, of, Code, The, Complete, Python, P...  
4  [React, The, Complete, Guide, 2023, incl, Reac...  
5  [The, Complete, JavaScript, Course, 2024, From...  
6  [Angular, The, Complete, Guide, 2023, Edition,...  
7  [Java, 17, Masterclass, Start, Coding, in, 202...  
8  [Machine, Learning, AZ, AI, Python, R, ChatGPT...  
9  [Python, for, Data, Science, and, Machine, Lea...

In [25]:
new_courses = courses[['id','title','tags']]

In [26]:
new_courses.head(10)

id                                              title  \
0  c00001  The Complete Python Bootcamp From Zero to Hero...   
1  c00002         The Complete 2023 Web Development Bootcamp   
2  c00003                    The Web Developer Bootcamp 2023   
3  c00004  100 Days of Code: The Complete Python Pro Boot...   
4  c00005  React - The Complete Guide 2023 (incl. React R...   
5  c00006  The Complete JavaScript Course 2024: From Zero...   
6  c00007        Angular - The Complete Guide (2023 Edition)   
7  c00008          Java 17 Masterclass: Start Coding in 2023   
8  c00009  Machine Learning A-Z: AI, Python & R + ChatGPT...   
9  c00010  Python for Data Science and Machine Learning B...   

                                                tags  
0  [The, Complete, Python, Bootcamp, From, Zero, ...  
1  [The, Complete, 2023, Web, Development, Bootca...  
2  [The, Web, Developer, Bootcamp, 2023, 10, Hour...  
3  [100, Days, of, Code, The, Complete, Python, P...  
4  [React, The, Complete, Guide, 2023, incl, Reac...  
5  [The, Complete, JavaScript, Course, 2024, From...  
6  [Angular, The, Complete, Guide, 2023, Edition,...  
7  [Java, 17, Masterclass, Start, Coding, in, 202...  
8  [Machine, Learning, AZ, AI, Python, R, ChatGPT...  
9  [Python, for, Data, Science, and, Machine, Lea...

In [27]:
# Converting tags from lists to strings
new_courses['tags'] = new_courses['tags'].apply(lambda x:" ".join(x))

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_22320\3918878773.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_courses['tags'] = new_courses['tags'].apply(lambda x:" ".join(x))


In [28]:
new_courses['tags'] = new_courses['tags'].str.lower()

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_22320\830239696.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_courses['tags'] = new_courses['tags'].str.lower()


In [29]:
new_courses.head(10)

id                                              title  \
0  c00001  The Complete Python Bootcamp From Zero to Hero...   
1  c00002         The Complete 2023 Web Development Bootcamp   
2  c00003                    The Web Developer Bootcamp 2023   
3  c00004  100 Days of Code: The Complete Python Pro Boot...   
4  c00005  React - The Complete Guide 2023 (incl. React R...   
5  c00006  The Complete JavaScript Course 2024: From Zero...   
6  c00007        Angular - The Complete Guide (2023 Edition)   
7  c00008          Java 17 Masterclass: Start Coding in 2023   
8  c00009  Machine Learning A-Z: AI, Python & R + ChatGPT...   
9  c00010  Python for Data Science and Machine Learning B...   

                                                tags  
0  the complete python bootcamp from zero to hero...  
1  the complete 2023 web development bootcamp bec...  
2  the web developer bootcamp 2023 10 hours of re...  
3  100 days of code the complete python pro bootc...  
4  react the complete guide 2023 incl react route...  
5  the complete javascript course 2024 from zero ...  
6  angular the complete guide 2023 edition master...  
7  java 17 masterclass start coding in 2023 acqui...  
8  machine learning az ai python r chatgpt prize ...  
9  python for data science and machine learning b...

In [30]:
#using nltk to convert the tags to root words to avoid wrong tags design like acting, acts, acted and keeping act instead
# import nltk
# from nltk.stem.porter import PorterStemmer
# ps = PorterStemmer()

In [31]:
#function to convert tag words to stem/root words
# def stem(text):
#     y = []
#     for i in text.split():
#         y.append(ps.stem(i))

#     return " ".join(y)

In [32]:
# new_courses['tags'] = new_courses['tags'].apply(stem)

In [33]:
new_courses

id                                              title  \
0     c00001  The Complete Python Bootcamp From Zero to Hero...   
1     c00002         The Complete 2023 Web Development Bootcamp   
2     c00003                    The Web Developer Bootcamp 2023   
3     c00004  100 Days of Code: The Complete Python Pro Boot...   
4     c00005  React - The Complete Guide 2023 (incl. React R...   
...      ...                                                ...   
3002  c03004  Complete JS Bootcamp | JavaScript Programming ...   
3003  c03005  Computer Vision: Face Recognition Quick Starte...   
3004  c03006             Statistics with R - Intermediate Level   
3005  c03007  Anime Character Creator: Make 3D Anime Charact...   
3006  c03008  Java Certification : OCP (1Z0-809) Exam Simula...   

                                                   tags  
0     the complete python bootcamp from zero to hero...  
1     the complete 2023 web development bootcamp bec...  
2     the web developer bootcamp 2023 10 hours of re...  
3     100 days of code the complete python pro bootc...  
4     react the complete guide 2023 incl react route...  
...                                                 ...  
3002  complete js bootcamp javascript programming in...  
3003  computer vision face recognition quick starter...  
3004  statistics with r intermediate level statistic...  
3005  anime character creator make 3d anime characte...  
3006  java certification ocp 1z0809 exam simulation ...  

[3007 rows x 3 columns]

In [34]:
#Using scikit learn module to use CountVectorizer for vectorisation
from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features = 2000,stop_words='english')

In [35]:
vectors = cv.fit_transform(new_courses['tags']).toarray()
vectors

array([[0, 0, 0, ..., 1, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [1, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [36]:
vectors.shape

(3007, 2000)

In [37]:
feature_names = cv.get_feature_names()
print(feature_names)

AttributeError: 'CountVectorizer' object has no attribute 'get_feature_names'

In [38]:
#using cosine similarity for finding distance between vectors
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vectors)

In [39]:
similarity.shape

(3007, 3007)

In [40]:
similarity[0]

array([1.        , 0.08703883, 0.04351941, ..., 0.        , 0.        ,
       0.        ])

In [41]:
sorted(similarity[1],reverse= True)

[0.9999999999999999,
 0.6363636363636364,
 0.6030226891555273,
 0.5938556868486718,
 0.568796458994521,
 0.5687964589945209,
 0.5543218625245872,
 0.5482304134858553,
 0.5360877067472385,
 0.5360877067472385,
 0.5222329678670936,
 0.5222329678670935,
 0.5146749286021821,
 0.511766315719159,
 0.5056498968474314,
 0.5029556907695452,
 0.5,
 0.4999999999999999,
 0.4994453686590201,
 0.4923659639173309,
 0.4923659639173309,
 0.4834937784152281,
 0.4787135538781691,
 0.4690415759823429,
 0.46848748060169054,
 0.4670993664969137,
 0.45133546692421994,
 0.444554224474387,
 0.44020438924006655,
 0.4281744192888376,
 0.42817441928883754,
 0.4264014327112209,
 0.41367015199601526,
 0.41030496993110904,
 0.40909090909090906,
 0.4021998332699218,
 0.3995562949272161,
 0.39886201760873286,
 0.39590379123244784,
 0.39167472590032015,
 0.39167472590032015,
 0.39167472590032015,
 0.38924947208076144,
 0.38924947208076144,
 0.38533731779422614,
 0.38533731779422614,
 0.38533731779422614,
 0.37630890450

In [42]:
sorted(list(enumerate(similarity[1])),reverse = True,key=lambda x:x[1])[1:6]

[(2, 0.6363636363636364),
 (677, 0.6030226891555273),
 (2230, 0.5938556868486718),
 (341, 0.568796458994521),
 (1752, 0.5687964589945209)]

In [43]:
new_courses.iloc[1].title

'The Complete 2023 Web Development Bootcamp'

In [44]:
#Final logic for recommendation
def recommed(title):
    course_index = new_courses[new_courses['title'] == title].index[0]
    distances = similarity[course_index]
    courses_list = sorted(list(enumerate(distances)),reverse = True,key=lambda x:x[1])[1:11]

    for i in courses_list:
            print(new_courses.iloc[i[0]].title)

In [45]:
recommed('GIT and Visual Studio with Azure DevOps Repos for Developers')

Learn Azure step by step.
Git with Visual Studio Code
Visual Studio Code for Developers 2024: Beginner to Advanced
Git & GitHub Complete Masterclass : Beginner to Git Expert
Azure Devops Crash Course -Build CI/CD release pipelines
Git for Geeks: Quick Git Training for Developers
Katalon Studio Beginners - Hands-on Training
Git by Example
Learn Visual Studio Code (2020)
Gitting Started: Step-by-Step Git and Github Crash Course
